Meteor early warning system

lessthnthree opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
Reintroduces an early warning system for the meteors random event. The station receives 3-5 minutes notice before the meteor wave is going to hit. This gives staff time to order meteor shields from cargo (if required) and set them up in anticipation, and overall the station is more prepared for the incoming wave.
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## How This Contributes To The Skyrat Roleplay Experience
- Oldbase had a meteor warning which was lost with TG, but for the pace of our server it's appropriate to give the crew time to prepare for a meteor wave, since we're on the station for 3 hours+ per shift.
<!-- Please add a short description of why you think these changes would benefit the game and the roleplay atmosphere of the server. If you can't justify it in words, it might not be worth adding. -->

## Proof of Testing

<!-- Include any screenshots/videos/debugging steps of the code functioning successfully, between the </summary> and </details> code blocks. -->
<!-- To our mappers and spriters: Posting screenshots of content INSIDE EDITORS (aseprite, PDN, SDMM, ect) is NOT valid proof of testing. Please make sure that you COMPILE the game and provide PROOF you tested your edits. -->

<details>
<summary>Screenshots/Videos</summary>

![image](https://user-images.githubusercontent.com/83487515/213347766-6e7c25e6-92c6-4d08-9703-35b0b75fba7e.png)

</details>

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl: LT3
add: New! Meteor wave early warning system. Prepare those meteor shields
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

OrionTheFox commented:
I have arrived from the other PR bearing news!

TG's directional framework applies to the "Sandstorm" event, which, as I've just learned, is NOT a subtype of meteors.
**But, it applies to the spawn_meteors proc. Bless.**
Sandstorm still serves as the perfect example of implementation though:
**_`/datum/round_event_control/sandstorm` in `code\modules\events\sandstorm.dm` has all the examples you should be copying off of._**

The variable `var/start_side`, selectable in `admin_setup()`, applies to the round_event_control. The same named variable is also added to `/datum/round_event/sandstorm`, and after checking to see if the event_control has one selected, is either set to that or randomly chosen, the announcement chooses a string for the direction, and then when the event fires, `spawn_meteors()` is called with the direction used in the third argument (`(number = 10, list/meteor_types, direction)`)

Overall not much more than using the already added meteors file (which might need renaming? idk) to modify the procs, or at least a place to comment that you've gone and modified the TG procs.

lessthnthree commented:
Okay I'll take a peek.

lessthnthree commented:
> I have arrived from the other PR bearing news!
> 
> TG's directional framework applies to the "Sandstorm" event, which, as I've just learned, is NOT a subtype of meteors. **But, it applies to the spawn_meteors proc. Bless.** Sandstorm still serves as the perfect example of implementation though: **_`/datum/round_event_control/sandstorm` in `code\modules\events\sandstorm.dm` has all the examples you should be copying off of._**
> 
> The variable `var/start_side`, selectable in `admin_setup()`, applies to the round_event_control. The same named variable is also added to `/datum/round_event/sandstorm`, and after checking to see if the event_control has one selected, is either set to that or randomly chosen, the announcement chooses a string for the direction, and then when the event fires, `spawn_meteors()` is called with the direction used in the third argument (`(number = 10, list/meteor_types, direction)`)
> 
> Overall not much more than using the already added meteors file (which might need renaming? idk) to modify the procs, or at least a place to comment that you've gone and modified the TG procs.

Done.

OrionTheFox commented on modular_skyrat/modules/new_meteors/code/new_meteors.dm:
```diff
@@ -0,0 +1,53 @@
+	else
+		start_side = pick(GLOB.cardinals)
+
+	priority_announce("Meteors have been detected on collision course with the station. The early warning system estimates first collision in approximately [start_when * 2] seconds, coming from the [dir2text(start_side)]. Ensure all sensitive areas and equipment are shielded.", "Meteor Alert", ANNOUNCER_METEORS)
```
For the sake of consistency with all other events like this, having the side be a ship direction rather than a cardinal direction would be best (Fore, Aft, Port, Starboard)

lessthnthree replied:
Sandstorm uses up/down/left/right?

lessthnthree replied:
Do we even have a proc that takes the integer and turns it into cardinal directions?

OrionTheFox replied:
Sandstorm's announcement reads as follows:
```
/datum/round_event/sandstorm/announce(fake)
	var/datum/round_event_control/sandstorm/sandstorm_event = control
	if(sandstorm_event.start_side)
		start_side = sandstorm_event.start_side
	else
		start_side = pick(GLOB.cardinals)

	var/start_side_text = "unknown"
	switch(start_side)
		if(NORTH)
			start_side_text = "fore"
		if(SOUTH)
			start_side_text = "aft"
		if(EAST)
			start_side_text = "starboard"
		if(WEST)
			start_side_text = "port"
		else
			stack_trace("Sandstorm event given [start_side] as unrecognized direction. Cancelling event...")
			kill()
			return

	priority_announce("A large wave of space dust is approaching from the [start_side_text] side of the station. \
	etc etc etc
```
it converts start side into start_side_text in relation to the nautical direction, rather than cardinal

I wish it was a proc. TG could probably use a proc for it...

lessthnthree replied:
I'll throw this in soon.

OrionTheFox commented on modular_skyrat/modules/new_meteors/code/new_meteors.dm:
```diff
@@ -0,0 +1,53 @@
+#define METEOR_WAVE_MIN_NOTICE 210
+#define METEOR_WAVE_MAX_NOTICE 240
+#define METEOR_WAVE_DURATION 37
```
Might want comments saying how long these are in real time rather than code time (I do not know bc I am stupid)

lessthnthree replied:
Sure

YakumoChen commented:
Why 7 minutes minimum? I'm fairly sure oldbase it was like, 3 to 5. We should reward engineers for doing preventive work more than just giving them a fairly luxurious amount of time to set things up.

lessthnthree commented:
> Why 7 minutes minimum? I'm fairly sure oldbase it was like, 3 to 5. We should reward engineers for doing preventive work more than just giving them a fairly luxurious amount of time to set things up only when meteors are announced.

I was told it was 5-10 on oldbase.

YakumoChen commented:
> > Why 7 minutes minimum? I'm fairly sure oldbase it was like, 3 to 5. We should reward engineers for doing preventive work more than just giving them a fairly luxurious amount of time to set things up only when meteors are announced.
> 
> I was told it was 5-10 on oldbase.

https://github.com/Skyrat-SS13/Skyrat13/pull/173

It was lowered for making the meteor event too trivial.

YakumoChen commented:
3-5 is a good balance. Not enough time for a ton of coverage if cargo is unable to order quickly, still enough time for a few emergency shield devices from secure storage to mitigate.

RatFromTheJungle commented:
> > > Why 7 minutes minimum? I'm fairly sure oldbase it was like, 3 to 5. We should reward engineers for doing preventive work more than just giving them a fairly luxurious amount of time to set things up only when meteors are announced.
> > 
> > 
> > I was told it was 5-10 on oldbase.
> 
> [Skyrat-SS13/Skyrat13#173](https://github.com/Skyrat-SS13/Skyrat13/pull/173)
> 
> It was lowered for making the meteor event too trivial.

(by citadel)

> If we make meteors so easy to deal with they're irrelevant every time, there's no point to the event.

every event doesn't need to be 'hard' to deal with, doubly so if it's going to randomly show up and level a fifth of the station
> Doubly so if they know the direction, since that saves them 3/4ths of the work and they can put up more shields.

doubly-so if we want to keep it in rotation, because the moment it starts routinely leveling parts of the station everyround-or-every-other it'd just be removed, or elsewise, gutted in effect

RatFromTheJungle commented:
> 3-5 is a good balance. Not enough time for a ton of coverage if cargo is unable to order quickly, still enough time for a few emergency shield devices from secure storage to mitigate. Rewards engineers taking initiative and gives crew time to evacuate.

3-5 isn't a good balance, 3-5 leaves, considering the cargo shuttle takes a minute to leave, and one to get back, probably a minute to unload if there's alot of a crates, then having to drag them out to space and around the station, about 30s to actually plant them around the station, assuming you're not on a multi-Z station
>  I can't recall the last time meteors beat the station down enough to be a round-ender.

essentially every time they hit atmos or engineering, is usually followed by a shuttlecall because no one wants to fix it, which is more an upstream issue of rebuilding being slow and tedious, but we work with what we have

YakumoChen commented:
> 3-5 isn't a good balance, 3-5 leaves, considering the cargo shuttle takes a minute to leave, and one to get back, probably a minute to unload if there's alot of a crates, then having to drag them out to space and around the station, about 30s to actually plant them around the station, assuming you're not on a multi-Z station

They can do them in advance, or just order the satellites early to have them ready.

> essentially every time they hit atmos or engineering, is usually followed by a shuttlecall because no one wants to fix it, which is more an upstream issue of rebuilding being slow and tedious, but we work with what we have

Directional makes this happen four times less often, and I almost never see shuttlecalls over meteors, usually only as an end point to an already messy round.

> oubly-so if we want to keep it in rotation, because the moment it starts routinely leveling parts of the station everyround-or-every-other it'd just be removed, or elsewise, gutted in effect

It's very rare already, the intent of the event is to give engineers more work.

We also have two different ERT buttons and many more options to fix the damage.

GoldenAlpharex commented on modular_skyrat/modules/new_meteors/code/new_meteors.dm:
```diff
@@ -0,0 +1,69 @@
+// 1 tick = 2 seconds
+#define METEOR_WAVE_MIN_NOTICE 210 // 7 minutes (420 seconds)
+#define METEOR_WAVE_MAX_NOTICE 270 // 9 minutes (540 seconds)
```
Could do this instead. You would then undef `IN_TICK` at the end of the file, obviously.
```suggestion
/// Each tick lasts two seconds.
#define IN_TICKS * 0.5
#define METEOR_WAVE_MIN_NOTICE 7 MINUTES IN_TICKS
#define METEOR_WAVE_MAX_NOTICE 9 MINUTES IN_TICKS
```

GoldenAlpharex commented on modular_skyrat/modules/new_meteors/code/new_meteors.dm:
```diff
@@ -0,0 +1,69 @@
+	var/start_side
+
+/datum/round_event/meteor_wave/New()
+	..()
```
```suggestion
	. = ..()
```

GoldenAlpharex commented on modular_skyrat/modules/new_meteors/code/new_meteors.dm:
```diff
@@ -0,0 +1,69 @@
+		return ADMIN_CANCEL_EVENT
+
+	var/force_dir = tgui_alert(usr, "Choose a side to throw rocks at?", "Interrupt some ERP?", list("Yes", "No", "Cancel"))
+	if(force_dir == "Cancel")
+		return ADMIN_CANCEL_EVENT
+
+	if(force_dir == "Yes")
+		var/chosen_direction = tgui_input_list(usr, "Pick one!","Plausible Deniability.", list("North", "South", "East", "West"))
+		switch(chosen_direction)
+			if("North")
+				start_side = NORTH
+			if("South")
+				start_side = SOUTH
+			if("East")
+				start_side = EAST
+			if("West")
+				start_side = WEST
```
Early returns are nice and simple, and this way you also properly cover the event that admins click the "X" button instead.
```suggestion
	if(force_dir != "Yes")
		return ADMIN_CANCEL_EVENT

	var/chosen_direction = tgui_input_list(usr, "Pick one!","Plausible Deniability.", list("North", "South", "East", "West"))
	switch(chosen_direction)
		if("North")
			start_side = NORTH
		if("South")
			start_side = SOUTH
		if("East")
			start_side = EAST
		if("West")
			start_side = WEST
```

GoldenAlpharex commented on modular_skyrat/modules/new_meteors/code/new_meteors.dm:
```diff
@@ -0,0 +1,69 @@
+
+/datum/round_event/meteor_wave/announce(fake)
+	var/datum/round_event_control/meteor_wave/meteor_event = control
+	if(meteor_event.start_side)
+		start_side = meteor_event.start_side
+	else
+		start_side = pick(GLOB.cardinals)
```
A much neater way to write this :)
```suggestion
	start_side = meteor_event.start_side || pick(GLOB.cardinals)
```

GoldenAlpharex commented on modular_skyrat/modules/new_meteors/code/new_meteors.dm:
```diff
@@ -0,0 +1,69 @@
+			return
+
+	priority_announce("Meteors have been detected on collision course with the station. The early warning system estimates first collision in approximately [start_when * 2] seconds, coming from the [start_side_text] side of the station. Ensure all sensitive areas and equipment are shielded.", "Meteor Alert", ANNOUNCER_METEORS)
+
```
```suggestion


```

GoldenAlpharex commented on modular_skyrat/modules/new_meteors/code/new_meteors.dm:
```diff
@@ -0,0 +1,69 @@
+	priority_announce("Meteors have been detected on collision course with the station. The early warning system estimates first collision in approximately [start_when * 2] seconds, coming from the [start_side_text] side of the station. Ensure all sensitive areas and equipment are shielded.", "Meteor Alert", ANNOUNCER_METEORS)
+
+/datum/round_event/meteor_wave/tick()
+	if(ISMULTIPLE(activeFor, 3))
```
That `3` should probably be a define.

GoldenAlpharex commented on modular_skyrat/modules/new_meteors/code/new_meteors.dm:
```diff
@@ -0,0 +1,69 @@
+
+/datum/round_event/meteor_wave/tick()
+	if(ISMULTIPLE(activeFor, 3))
+		spawn_meteors(5, wave_type, start_side) //meteor list types defined in gamemode/meteor/meteors.dm
```
This `5` should probably be a define too.

Quink-Fix commented:
I would suggest changing 'North', 'East', 'South' and 'West' to their nautical counterparts (assuming the Bridge is 'North', it'd be 'Bow', 'Starboard Side', 'Stern', and 'Port Side') to match the naming conventions in the rest of the game. They are used in area names which are used for many announcements, so there's no reason not to use them here.

lessthnthree commented:
> I would suggest changing 'North', 'East', 'South' and 'West' to their nautical counterparts (assuming the Bridge is 'North', it'd be 'Bow', 'Starboard Side', 'Stern', and 'Port Side') to match the naming conventions in the rest of the game. They are used in area names which are used for many announcements, so there's no reason not to use them here.

It does, though not reflected in the screenshot.

OrionTheFox commented on modular_skyrat/modules/new_meteors/code/new_meteors.dm:
```diff
@@ -0,0 +1,78 @@
+	var/datum/round_event_control/meteor_wave/meteor_event = control
+	start_side = meteor_event.start_side || pick(GLOB.cardinals)
+
+	var/start_side_text = "unknown"
```
Did we ever get an update on the PR to make a proc for nautical directions?

lessthnthree replied:
Yeah it died at TG. Maybe after freeze.

OrionTheFox commented on modular_skyrat/modules/new_meteors/code/new_meteors.dm:
```diff
@@ -0,0 +1,78 @@
+		if(WEST)
+			start_side_text = "port"
+		else
+			stack_trace("Sandstorm event given [start_side] as unrecognized direction. Cancelling event...")
```
```suggestion
			stack_trace("Meteor event given [start_side] as unrecognized direction. Cancelling event...")
```

OrionTheFox commented on modular_skyrat/modules/new_meteors/code/new_meteors.dm:
```diff
@@ -0,0 +1,78 @@
+			kill()
+			return
+
+	priority_announce("Meteors have been detected on collision course with the station. The early warning system estimates first collision in approximately [start_when * 2] seconds, coming from the [start_side_text] side of the station. Ensure all sensitive areas and equipment are shielded.", "Meteor Alert", ANNOUNCER_METEORS)
```
```suggestion
	priority_announce("Meteors have been detected on collision course with the station. Estimated collision in approximately [start_when * 2] seconds from the [start_side_text]-side of the station. Ensure all sensitive areas and equipment are shielded.", "Meteor Alert", ANNOUNCER_METEORS)
```
Bit more concise, not necessary but less wordy is good for emergencies

OrionTheFox commented on modular_skyrat/modules/new_meteors/readme.md:
```diff
@@ -0,0 +1,16 @@
+https://github.com/Skyrat-SS13/Skyrat-tg/pull/18813
+
+## Title:
+MODULE ID: NEW_METEORS
```
sometimes i wish we had sub-modules, i.e. a big "Events" folder where we put all this stuff... woe is me...

lessthnthree commented:
DNM until https://github.com/Skyrat-SS13/Skyrat-tg/pull/19068 is merged.

Quink-Fix commented:
DNM should be cleared - unless this is blocked by the feature freeze?

Gandalf2k15 reviewed (changes requested):
Remove the directional feature, make the timer 3-5 minutes.

lessthnthree commented:
> Remove the directional feature, make the timer 3-5 minutes.

Done.
